### DISCUSSION TITLE
Using custom nbt for a model predicate provider.

### DISCUSSION THREAD

**Discussion by Desynq** [FabricMC/fabric-api#2366] [Mod Dev Support] [answered]
I'm currently trying to make an item that can store experience display a different model depending on how full it is of experience.

The experience is stored in them item under the an integer nbt tag known as `ExperienceStored`. I have an arbitrary system that restricts the player from being able to fill the experience crystal to more than 3000 experience points.

There are going to be 3 different models used, one for when it is under 33% full, one for between 33% and 66% full, and one for 66% or greater full.

Looking at Mojang's implementation of the `ModelPredicateProviderRegistry`, items such as bows can have their values translated as a clamped float between 0 and 1.

I've tried implementing said code while also using `ItemStack`'s `getOrCreateNbt` method for getting the amount of experience stored.

Here is the current code:
```java
ModelPredicateProviderRegistry.register(ModItems.EXPERIENCE_CRYSTAL, new Identifier("fill"), (stack, world, entity, seed) -> {
	if (entity == null) {
		return 0.0f;
	}
	if (entity.getActiveItem() != stack) {
		return 0.0f;
	}
	return (float) stack.getOrCreateNbt().getInt("StoredExperience") / 3000.0f;
});
```
This is within a `ModModelPredicateProviderRegistry` class that is initialized by my `ClientModInitializer` class.
I've also made sure to add the `ClientModInitializer` to the `fabric.mod.json` file entry points compound under `"client"`.

It currently does not work and I'm wondering if I can get some feedback or help on how `ModelPredicateProviderRegistry` works, specifically with NBT values.

**Comment by warjort** [comment, marked as answer]
I have never used this feature myself, but a quick search on github shows this small mod using it.
https://github.com/AmyMialeeMods/icy-incitement/blob/main/src/main/java/amymialee/icyincitement/IcyIncitementClient.java

Here's what the model looks like which seems to be similar to what you want?
https://github.com/AmyMialeeMods/icy-incitement/blob/main/src/main/resources/assets/icyincitement/models/item/snowball_sprinkler.json

BTW I think your 2 if statements are not relevant for your use case? In the code you copied, they represent forcing the uncharged bow to be drawn if there is no entity actively holding it.

If you are interested, it also shows how to draw a filled bar next to the item using a mixin:
https://github.com/AmyMialeeMods/icy-incitement/blob/main/src/main/java/amymialee/icyincitement/mixin/ItemRendererMixin.java

**Comment by Daomephsta** [comment]
Sidenote:  
`new Identifier("fill")` is equivalent to `new Identifier("minecraft:fill")`.
When creating identifiers for your mod, you must always explicitly specify your mod's namespace;   
i.e. `new Identifier("examplemod:fill")` or `new Identifier("examplemod", "fill")`. 

Using the `minecraft` namespace has 2 issues:
1. Content appears to be vanilla when it isn't
2. Conflicts with other mods that also use the `minecraft` namespace
